Fail gracefully if the application is not allowed to write files #73

ernestlim8 opened the pull request:
Fixes #73 

After a command, AB3 displays an error message and exits after three
seconds.

Automatically exiting the application is more user-friendly because the
user is unable to use the application purposefully with the file write
error.

Let's check for write permissions, then display the error message to the user showing
the file that has no permissions. The application will automatically
exit gracefully after three seconds.

MightyCupcakes commented:
Hello,

Addressbook is currently following the PR workflow outlined in [OSS generic](https://github.com/oss-generic/process/blob/master/docs/QuestionsIssuesPrs.md#submitting-a-pr)

In particular:

* Please only fix one issue per PR. If you want to work on multiple issues, open separate PRs for each issue.
* PR title should be the issue title including its identification number

ernestlim8 committed:
Improve error message if no write permission

When there is no write permission for the data file,
an AccessDeniedException will be shown in the error message
to the user.

Let's make the error message more user friendly by informing
the user that there are insufficient permissions for the file.

ernestlim8 commented:
Fixes #73

damithc commented:
> Fixes #73

This should be in the first post for GitHub to close the issue when this PR is merged.

JinHao-L commented on src/main/java/seedu/address/ui/MainWindow.java:
```diff
@@ -188,9 +192,37 @@ private CommandResult executeCommand(String commandText) throws CommandException
+            }
+        };
+
+        new Thread(accessDeniedTask).start();
```
I found another alternative solution which is the [`PauseTransition`](https://docs.oracle.com/javase/8/javafx/api/javafx/animation/PauseTransition.html) class in the JavaFX library.

The following code produces the same effect and similarly does not freeze the UI. 
```
PauseTransition pause = new PauseTransition(Duration.seconds(3));
pause.setOnFinished(event -> handleExit());
pause.play();
```

I am not sure if this is more efficient but I feel that this solution might be more reader-friendly. 
What do you all think?

ernestlim8 replied:
Agreed, this solution seems more user friendly. It works on my end as well.

JinHao-L commented on src/main/java/seedu/address/ui/MainWindow.java:
```diff
@@ -188,9 +192,37 @@ private CommandResult executeCommand(String commandText) throws CommandException
+
+    private void handleAccessDeniedException(Exception e) {
+        logger.info("Insufficient Permissions for file");
+        resultDisplay.setFeedbackToUser(e.getMessage());
```
Currently, the feedback to the user seems like an unchecked exception:
> Could not save data to file: java.nio.file.AccessDeniedException: data\addressbook.json

Maybe it would be better if we rephrase it to `"Could not save data to file: Access Denied"` or `"Could not save data to file: Insufficient permission"`.

ernestlim8 replied:
I have updated the error message to be more readable for the user as per your suggestion.

`Could not save data to file: Insufficient permission for <filepath>`
The additional information of the file path allows the user to know which file path has insufficient permissions.

JinHao-L reviewed (commented):
Other than the comments below, the rest looks good! 👍

damithc reviewed (commented):
Thanks for the PR @ernestlim8 
I think it is enough to show a clear error message and stop operations. 
There is no need to exit automatically. Auto-exiting will make test automation harder (although we are not doing auto-testing of the GUI in AB3, it is done in AB4). Even in manual testing, it will make it harder for students to take screenshots if the error happens.

ernestlim8 commented:
@damithc After commenting out the Pause Transition code, if there are insufficient file permissions, AB3 will display the error message for all valid commands. Is this behaviour sufficient for stopping operations?

damithc commented:
> @damithc After commenting out the Pause Transition code, if there are insufficient file permissions, AB3 will display the error message for all valid commands. Is this behaviour sufficient for stopping operations?

Sorry @ernestlim8 , overlooked this until now due to start-of-semester rush. Yes, I think that would work. Is the current code doing just that? I'm asking because I saw some unusual import statements.

damithc commented:
@se-edu/tech-team-level1 for your review

damithc commented:
Sorry @ernestlim8 I think this PR will progress very slowly as this is the semester time and this change is not urgent.

damithc commented on src/main/java/seedu/address/ui/MainWindow.java:
```diff
@@ -188,9 +190,19 @@ private CommandResult executeCommand(String commandText) throws CommandException
+            } else {
+                resultDisplay.setFeedbackToUser(e.getMessage());
+                logger.info("Invalid command: " + commandText);
+            }
```
See if you can get rid of this condition (checking of object type should be avoided if possible). That is, if the correct error message was put into the exception in the first place, you don't need a different behaviors here. That means we only need to change the place where the error is detected.

damithc commented on src/main/java/seedu/address/ui/MainWindow.java:
```diff
@@ -188,8 +189,8 @@ private CommandResult executeCommand(String commandText) throws CommandException
         } catch (CommandException | ParseException e) {
-            logger.info("Invalid command: " + commandText);
             resultDisplay.setFeedbackToUser(e.getMessage());
+            logger.info("Invalid command: " + commandText);
```
Any reason for this change?

damithc commented on src/main/java/seedu/address/ui/MainWindow.java:
```diff
@@ -156,6 +156,7 @@ void show() {
      */
     @FXML
     private void handleExit() {
+        logger.info("Calling exit");
```
Seems this change is not related to this PR?

damithc reviewed (commented):
Sorry for getting back so late @ernestlim8 
Some minor comments added.

damithc commented:
@ernestlim8 shall we finish this PR?

damithc commented on src/main/java/seedu/address/logic/LogicManager.java:
```diff
@@ -48,7 +48,8 @@ public CommandResult execute(String commandText) throws CommandException, ParseE
         } catch (IOException ioe) {
-            throw new CommandException(FILE_OPS_ERROR_MESSAGE + ioe, ioe);
+            String message = ioe.getMessage();
+            throw new CommandException(String.format(FILE_OPS_ERROR_MESSAGE, message), ioe);
```
Looks like the code assumes every IOException is caused by _insufficient permissions_? I don't think that is a valid assumption.

ernestlim8 replied:
@damithc Agreed, I will display the insufficient permissions when catching the AccessDeniedException instead, and use a more generalised message for other possible I/O Exceptions.

damithc commented on src/main/java/seedu/address/ui/MainWindow.java:
```diff
@@ -188,7 +188,6 @@ private CommandResult executeCommand(String commandText) throws CommandException
 
             return commandResult;
         } catch (CommandException | ParseException e) {
-            logger.info("Invalid command: " + commandText);
```
What if the command is actually invalid?

ernestlim8 replied:
@damithc Seems like this log statement was already existing before this PR, its possible to split the 2 different exceptions and handle them separately, if the desired behaviour is to differentiate between invalid command and parse exception in the logs

damithc reviewed (commented):
I have some concerns about the logic of this fix. Happy to be proven wrong though.

damithc commented:
To be continued by #172 
Thanks for this PR @ernestlim8
